README.md
# @feathersjs/errors

[![Greenkeeper badge](https://badges.greenkeeper.io/feathersjs/errors.svg)](https://greenkeeper.io/)

[![Build Status](https://travis-ci.org/feathersjs/errors.png?branch=master)](https://travis-ci.org/feathersjs/errors)
[![Test Coverage](https://codeclimate.com/github/feathersjs/errors/badges/coverage.svg)](https://codeclimate.com/github/feathersjs/errors/coverage)
[![Dependency Status](https://img.shields.io/david/feathersjs/errors.svg?style=flat-square)](https://david-dm.org/feathersjs/errors)
[![Download Status](https://img.shields.io/npm/dm/@feathersjs/errors.svg?style=flat-square)](https://www.npmjs.com/package/@feathersjs/errors)

> Common error types for feathers apps

## Getting Started

Feathers errors come with feathers by default. So typically you don't need to install it at all.

In the event that you do need to install it:

```bash
npm install --save @feathersjs/errors
```

## Documentation

#### Current Error Types:

* `BadRequest`: 400
* `NotAuthenticated`: 401
* `PaymentError`: 402
* `Forbidden`: 403
* `NotFound`: 404
* `MethodNotAllowed`: 405
* `NotAcceptable`: 406
* `Timeout`: 408
* `Conflict`: 409
* `LengthRequired`: 411
* `Unprocessable`: 422
* `TooManyRequests`: 429
* `GeneralError`: 500
* `NotImplemented`: 501
* `BadGateway`: 502
* `Unavailable`: 503

**Pro Tip:** Feathers service adapters (ie. mongodb, memory, etc.) already emit the appropriate errors for you. :-)

#### Usage:

```js
import errors from '@feathersjs/errors';

// If you were to create an error yourself.
var notFound = new errors.NotFound('User does not exist');

// You can wrap existing errors
var existing = new errors.GeneralError(new Error('I exist'));

// You can also pass additional data
var data = new errors.BadRequest('Invalid email', {email: 'sergey@google.com'});

// You can also pass additional data
var dataWithoutMessage = new errors.BadRequest({email: 'sergey@google.com'});

// If you need to pass multiple errors
var validationErrors = new errors.BadRequest('Invalid Parameters', {errors: {email: 'Email already taken'} });

// You can also omit the error message and we'll put in a default one for you
var validationErrors = new errors.BadRequest({errors: {email: 'Invalid Email'} });
```

## License

Copyright (c) 2016 Feathers Contributors

Licensed under the [MIT license](LICENSE).


CHANGELOG.md
# Change Log

## [v2.9.2](https://github.com/feathersjs/errors/tree/v2.9.2) (2017-09-05)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.9.1...v2.9.2)

**Closed issues:**

- Getting 500 status code when attempting to throw non 500 style errors \(401\) [\#85](https://github.com/feathersjs/errors/issues/85)

**Merged pull requests:**

- fix typings [\#87](https://github.com/feathersjs/errors/pull/87) ([j2L4e](https://github.com/j2L4e))
- Update debug to the latest version 🚀 [\#86](https://github.com/feathersjs/errors/pull/86) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))
- Update sinon to the latest version 🚀 [\#84](https://github.com/feathersjs/errors/pull/84) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))

## [v2.9.1](https://github.com/feathersjs/errors/tree/v2.9.1) (2017-07-21)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.9.0...v2.9.1)

**Merged pull requests:**

- Add back default error message [\#83](https://github.com/feathersjs/errors/pull/83) ([daffl](https://github.com/daffl))

## [v2.9.0](https://github.com/feathersjs/errors/tree/v2.9.0) (2017-07-20)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.8.2...v2.9.0)

**Closed issues:**

- Wrong stack for errors [\#78](https://github.com/feathersjs/errors/issues/78)

**Merged pull requests:**

- Capture proper stack trace and error messages [\#82](https://github.com/feathersjs/errors/pull/82) ([daffl](https://github.com/daffl))
- Update chai to the latest version 🚀 [\#81](https://github.com/feathersjs/errors/pull/81) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))

## [v2.8.2](https://github.com/feathersjs/errors/tree/v2.8.2) (2017-07-05)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.8.1...v2.8.2)

**Merged pull requests:**

- Fix wildcard import on ES2015+ [\#80](https://github.com/feathersjs/errors/pull/80) ([coreh](https://github.com/coreh))
- Add more information to error debug [\#79](https://github.com/feathersjs/errors/pull/79) ([kamzil](https://github.com/kamzil))

## [v2.8.1](https://github.com/feathersjs/errors/tree/v2.8.1) (2017-05-30)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.8.0...v2.8.1)

**Merged pull requests:**

- Fix errors property being lost when cloning [\#76](https://github.com/feathersjs/errors/pull/76) ([daniel-constantin](https://github.com/daniel-constantin))

## [v2.8.0](https://github.com/feathersjs/errors/tree/v2.8.0) (2017-05-08)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.7.1...v2.8.0)

**Closed issues:**

- Support array objects as data [\#64](https://github.com/feathersjs/errors/issues/64)

**Merged pull requests:**

- Allow data to be an array [\#75](https://github.com/feathersjs/errors/pull/75) ([daniel-constantin](https://github.com/daniel-constantin))

## [v2.7.1](https://github.com/feathersjs/errors/tree/v2.7.1) (2017-04-28)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.7.0...v2.7.1)

**Closed issues:**

- Object.setPrototypeOf in IE 10 [\#70](https://github.com/feathersjs/errors/issues/70)

**Merged pull requests:**

- Define property toJSON because just assigning it throws an error in N… [\#74](https://github.com/feathersjs/errors/pull/74) ([daffl](https://github.com/daffl))

## [v2.7.0](https://github.com/feathersjs/errors/tree/v2.7.0) (2017-04-25)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.6.3...v2.7.0)

**Merged pull requests:**

- Change back to old Error inheritance [\#73](https://github.com/feathersjs/errors/pull/73) ([daffl](https://github.com/daffl))
- Update semistandard to the latest version 🚀 [\#72](https://github.com/feathersjs/errors/pull/72) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))
- Update dependencies to enable Greenkeeper 🌴 [\#71](https://github.com/feathersjs/errors/pull/71) ([greenkeeper[bot]](https://github.com/apps/greenkeeper))

## [v2.6.3](https://github.com/feathersjs/errors/tree/v2.6.3) (2017-04-08)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.6.2...v2.6.3)

**Closed issues:**

- Make options the same as res.format [\#37](https://github.com/feathersjs/errors/issues/37)

**Merged pull requests:**

- fix typescript definitions with noImplicitAny [\#69](https://github.com/feathersjs/errors/pull/69) ([JVirant](https://github.com/JVirant))

## [v2.6.2](https://github.com/feathersjs/errors/tree/v2.6.2) (2017-03-16)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.6.1...v2.6.2)

**Closed issues:**

- Create a TokenExpired error type [\#53](https://github.com/feathersjs/errors/issues/53)

**Merged pull requests:**

- Fix declarations for index.d.ts [\#66](https://github.com/feathersjs/errors/pull/66) ([ghost](https://github.com/ghost))

## [v2.6.1](https://github.com/feathersjs/errors/tree/v2.6.1) (2017-03-06)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.6.0...v2.6.1)

**Merged pull requests:**

- fix pull request \#62 [\#63](https://github.com/feathersjs/errors/pull/63) ([superbarne](https://github.com/superbarne))

## [v2.6.0](https://github.com/feathersjs/errors/tree/v2.6.0) (2017-03-04)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.5.0...v2.6.0)

**Closed issues:**

- Full Validation Error Object not passed to client promise [\#61](https://github.com/feathersjs/errors/issues/61)
- More HTTP Statuses [\#48](https://github.com/feathersjs/errors/issues/48)

**Merged pull requests:**

- add typescript definitions [\#62](https://github.com/feathersjs/errors/pull/62) ([superbarne](https://github.com/superbarne))
- Fix compile npm task on Windows [\#60](https://github.com/feathersjs/errors/pull/60) ([AbraaoAlves](https://github.com/AbraaoAlves))

## [v2.5.0](https://github.com/feathersjs/errors/tree/v2.5.0) (2016-11-04)
[Full Changelog](https://github.com/feathersjs/errors/compare/v2.4.0...v2.5.0)

**Closed issues:**

- Possible issue with Node 4 [\#51](https://github.com/feathersjs/errors/issues/51)

test/index.test.js
import assert from 'assert';
import { types, errors, convert } from '../src';

describe('feathers-errors', () => {
  it('is CommonJS compatible', () => {
    assert.equal(typeof require('../lib'), 'object');
    assert.equal(typeof require('../lib').FeathersError, 'function');
  });

  describe('errors.convert', () => {
    it('converts objects to feathers errors', () => {
      const error = convert({
        name: 'BadRequest',
        message: 'Hi',
        expando: 'Me'
      });

      assert.ok(error instanceof errors.BadRequest);
      assert.equal(error.message, 'Hi');
      assert.equal(error.expando, 'Me');
    });

    it('converts other object to error', () => {
      let error = convert({
        message: 'Something went wrong'
      });

      assert.ok(error instanceof Error);
      assert.equal(error.message, 'Something went wrong');

      error = convert('Something went wrong');

      assert.ok(error instanceof Error);
      assert.equal(error.message, 'Something went wrong');
    });
  });

  describe('error types', () => {
    it('Bad Request', () => {
      assert.notEqual(typeof errors.BadRequest, 'undefined', 'has BadRequest');
    });

    it('Not Authenticated', () => {
      assert.notEqual(typeof errors.NotAuthenticated, 'undefined', 'has NotAuthenticated');
    });

    it('Payment Error', () => {
      assert.notEqual(typeof errors.PaymentError, 'undefined', 'has PaymentError');
    });

    it('Forbidden', () => {
      assert.notEqual(typeof errors.Forbidden, 'undefined', 'has Forbidden');
    });

    it('Not Found', () => {
      assert.notEqual(typeof errors.NotFound, 'undefined', 'has NotFound');
    });

    it('Method Not Allowed', () => {
      assert.notEqual(typeof errors.MethodNotAllowed, 'undefined', 'has MethodNotAllowed');
    });

    it('Not Acceptable', () => {
      assert.notEqual(typeof errors.NotAcceptable, 'undefined', 'has NotAcceptable');
    });

    it('Timeout', () => {
      assert.notEqual(typeof errors.Timeout, 'undefined', 'has Timeout');
    });

    it('Conflict', () => {
      assert.notEqual(typeof errors.Conflict, 'undefined', 'has Conflict');
    });

    it('Length Required', () => {
      assert.notEqual(typeof errors.LengthRequired, 'undefined', 'has LengthRequired');
    });

    it('Unprocessable', () => {
      assert.notEqual(typeof errors.Unprocessable, 'undefined', 'has Unprocessable');
    });

    it('Too Many Requests', () => {
      assert.notEqual(typeof errors.TooManyRequests, 'undefined', 'has TooManyRequests');
    });

    it('General Error', () => {
      assert.notEqual(typeof errors.GeneralError, 'undefined', 'has GeneralError');
    });

    it('Not Implemented', () => {
      assert.notEqual(typeof errors.NotImplemented, 'undefined', 'has NotImplemented');
    });

    it('Bad Gateway', () => {
      assert.notEqual(typeof errors.BadGateway, 'undefined', 'has BadGateway');
    });

    it('Unavailable', () => {
      assert.notEqual(typeof errors.Unavailable, 'undefined', 'has Unavailable');
    });

    it('400', () => {
      assert.notEqual(typeof errors[400], 'undefined', 'has BadRequest alias');
    });

    it('401', () => {
      assert.notEqual(typeof errors[401], 'undefined', 'has NotAuthenticated alias');
    });

    it('402', () => {
      assert.notEqual(typeof errors[402], 'undefined', 'has PaymentError alias');
    });

    it('403', () => {
      assert.notEqual(typeof errors[403], 'undefined', 'has Forbidden alias');
    });

    it('404', () => {
      assert.notEqual(typeof errors[404], 'undefined', 'has NotFound alias');
    });

    it('405', () => {
      assert.notEqual(typeof errors[405], 'undefined', 'has MethodNotAllowed alias');
    });

    it('406', () => {
      assert.notEqual(typeof errors[406], 'undefined', 'has NotAcceptable alias');
    });

    it('408', () => {
      assert.notEqual(typeof errors[408], 'undefined', 'has Timeout alias');
    });

    it('409', () => {
      assert.notEqual(typeof errors[409], 'undefined', 'has Conflict alias');
    });

    it('411', () => {
      assert.notEqual(typeof errors[411], 'undefined', 'has LengthRequired alias');
    });

    it('422', () => {
      assert.notEqual(typeof errors[422], 'undefined', 'has Unprocessable alias');
    });

    it('429', () => {
      assert.notEqual(typeof errors[429], 'undefined', 'has TooManyRequests alias');
    });

    it('500', () => {
      assert.notEqual(typeof errors[500], 'undefined', 'has GeneralError alias');
    });

    it('501', () => {
      assert.notEqual(typeof errors[501], 'undefined', 'has NotImplemented alias');
    });

    it('502', () => {
      assert.notEqual(typeof errors[502], 'undefined', 'has BadGateway alias');
    });

    it('503', () => {
      assert.notEqual(typeof errors[503], 'undefined', 'has Unavailable alias');
    });

    it('instanatiates every error', () => {
      Object.keys(types).forEach(name => {
        const E = types[name];
        if (E) {
          new E('Something went wrong'); // eslint-disable-line no-new
        }
      });
    });
  });

  it('exposes errors via types for backwards compatibility', () => {
    assert.notEqual(typeof types.BadRequest, 'undefined', 'has BadRequest');
  });

  describe('successful error creation', () => {
    describe('without custom message', () => {
      it('default error', () => {
        var error = new errors.GeneralError();
        assert.equal(error.code, 500);
        assert.equal(error.className, 'general-error');
        assert.notEqual(error.stack, undefined);
        assert.equal(error instanceof errors.GeneralError, true);
        assert.equal(error instanceof errors.FeathersError, true);
      });

      it('can wrap an existing error', () => {
        var error = new errors.BadRequest(new Error());
        assert.equal(error.code, 400);
        assert.equal(error.message, 'Error');
      });

      it('with multiple errors', () => {
        var data = {
          errors: {
            email: 'Email Taken',
            password: 'Invalid Password'
          },
          foo: 'bar'
        };

        var error = new errors.BadRequest(data);
        assert.equal(error.code, 400);
        assert.equal(error.message, 'Error');
        assert.deepEqual(error.errors, {email: 'Email Taken', password: 'Invalid Password'});
        assert.deepEqual(error.data, {foo: 'bar'});
      });

      it('with data', () => {
        var data = {
          email: 'Email Taken',
          password: 'Invalid Password'
        };

        var error = new errors.GeneralError(data);
        assert.equal(error.code, 500);
        assert.equal(error.message, 'Error');
        assert.deepEqual(error.data, data);
      });
    });

    describe('with custom message', () => {
      it('contains our message', () => {
        var error = new errors.BadRequest('Invalid Password');
        assert.equal(error.code, 400);
        assert.equal(error.message, 'Invalid Password');
      });

      it('can wrap an existing error', () => {
        var error = new errors.BadRequest(new Error('Invalid Password'));
        assert.equal(error.code, 400);
        assert.equal(error.message, 'Invalid Password');
      });

      it('with data', () => {
        var data = {
          email: 'Email Taken',
          password: 'Invalid Password'
        };

        var error = new errors.GeneralError('Custom Error', data);
        assert.equal(error.code, 500);
        assert.equal(error.message, 'Custom Error');
        assert.deepEqual(error.data, data);
      });

      it('with multiple errors', () => {
        var data = {
          errors: {
            email: 'Email Taken',
            password: 'Invalid Password'
          },
          foo: 'bar'
        };

        var error = new errors.BadRequest(data);
        assert.equal(error.code, 400);
        assert.equal(error.message, 'Error');
        assert.deepEqual(error.errors, {email: 'Email Taken', password: 'Invalid Password'});
        assert.deepEqual(error.data, {foo: 'bar'});
      });
    });

    it('can return JSON', () => {
      var data = {
        errors: {
          email: 'Email Taken',
          password: 'Invalid Password'
        },
        foo: 'bar'
      };

      var expected = '{"name":"GeneralError","message":"Custom Error","code":500,"className":"general-error","data":{"foo":"bar"},"errors":{"email":"Email Taken","password":"Invalid Password"}}';

      var error = new errors.GeneralError('Custom Error', data);
      assert.equal(JSON.stringify(error), expected);
    });

    it('can handle immutable data', () => {
      var data = {
        errors: {
          email: 'Email Taken',
          password: 'Invalid Password'
        },
        foo: 'bar'
      };

      var error = new errors.GeneralError('Custom Error', Object.freeze(data));
      assert.equal(error.data.errors, undefined);
      assert.deepEqual(error.data, {foo: 'bar'});
    });

    it('allows arrays as data', () => {
      var data = [
        {
          hello: 'world'
        }
      ];
      data.errors = 'Invalid input';

      var error = new errors.GeneralError('Custom Error', data);
      assert.equal(error.data.errors, undefined);
      assert.ok(Array.isArray(error.data));
      assert.deepEqual(error.data, [{hello: 'world'}]);
      assert.equal(error.errors, 'Invalid input');
    });

    it('has proper stack trace (#78)', () => {
      try {
        throw new errors.NotFound('Not the error you are looking for');
      } catch (e) {
        const text = 'NotFound: Not the error you are looking for';

        assert.equal(e.stack.indexOf(text), 0);
        assert.ok(e.stack.indexOf('index.test.js') !== -1);
      }
    });
  });
});


test/errors.test.js
'use strict';

/* jshint undef:false */

var assert = require('assert');
var feathers = require('feathers');
var errors = require('../lib/errors');
var app;

describe('Feathers errors', function () {

  beforeEach(function() {
    app = feathers().configure(errors());
  });

  it('exposes the api properly', function () {
    assert.equal(typeof errors.types, 'object', 'exposes error types');
    assert.equal(typeof errors.handler, 'function', 'exposes error handler');
    assert.equal(typeof errors.missing, 'function', 'exposes 404 handler');
  });

  it('initializes errors object', function () {
    assert.equal(typeof app.errors, 'object', 'errors got added to the app');
  });

  it('can create app errors', function () {
    var error = new app.errors.GeneralError('foo');

    assert.equal(error.code, 500 ,'a general error was created with the correct code');
    assert.equal(error.message, 'foo' ,'a general error was created with the correct message');
  });

  it('app has all the available errors', function () {
    assert.notEqual(typeof app.errors.BadRequest, 'undefined', 'has BadRequest');
    assert.notEqual(typeof app.errors.NotAuthenticated, 'undefined', 'has NotAuthenticated');
    assert.notEqual(typeof app.errors.PaymentError, 'undefined', 'has PaymentError');
    assert.notEqual(typeof app.errors.Forbidden, 'undefined', 'has Forbidden');
    assert.notEqual(typeof app.errors.NotFound, 'undefined', 'has NotFound');
    assert.notEqual(typeof app.errors.MethodNotAllowed, 'undefined', 'has MethodNotAllowed');
    assert.notEqual(typeof app.errors.NotAcceptable, 'undefined', 'has NotAcceptable');
    assert.notEqual(typeof app.errors.Timeout, 'undefined', 'has Timeout');
    assert.notEqual(typeof app.errors.Conflict, 'undefined', 'has Conflict');
    assert.notEqual(typeof app.errors.Unprocessable, 'undefined', 'has Unprocessable');
    assert.notEqual(typeof app.errors.GeneralError, 'undefined', 'has GeneralError');
    assert.notEqual(typeof app.errors.NotImplemented, 'undefined', 'has NotImplemented');
    assert.notEqual(typeof app.errors.Unavailable, 'undefined', 'has Unavailable');
  });

  it.skip('error handler catches errors', function () {
    // TODO (EK)
  });
});


test/handler.test.js
if(!global._babelPolyfill) { require('babel-polyfill'); }

import feathers from 'feathers';
import assert from 'assert';
import request from 'request';
import fs from 'fs';
import { join } from 'path';
import { errors } from '../src';
import handler from '../src/handler';

describe('feathers-errors', () => {
  it('is CommonJS compatible', () => {
    assert.equal(typeof require('../lib/handler'), 'function');
  });

  it('is import compatible', () => {
    assert.equal(typeof handler, 'function');
  });
  
  describe('use as app error handler', function() {
    before(function() {
      this.app = feathers()
        .get('/error', function(req, res, next) {
          next(new Error('Something went wrong'));
        })
        .get('/string-error', function(req, res, next) {
          const e = new Error('Something was not found');
          e.code = '404';
          
          next(e);
        })
        .use(function(req, res, next) {
          next(new errors.NotFound('File not found'));
        })
        .use(handler());
      
      this.server = this.app.listen(5050);
    });
    
    after(function(done) {
      this.server.close(done);
    });
    
    describe('converts an non-feathers error', () => {
      it('is an instance of GeneralError', done => {
        request({
          url: 'http://localhost:5050/error',
          json: true
        }, (error, res, body) => {
          assert.equal(res.statusCode, 500);
          assert.deepEqual(body, {
            name: 'GeneralError',
            message: 'Something went wrong',
            code: 500,
            className: 'general-error',
            data: {},
            errors: {}
          });
          done();
        });
      });

      it.skip('still has a stack trace', () => {
        assert.equal(handler, 'function');  
      });
    });

    describe('text/html format', () => {
      it('serves a 404.html', done => {
        fs.readFile(join(__dirname, '..', 'src', 'public', '404.html'), function(err, html) {
          request('http://localhost:5050/path/to/nowhere', (error, res, body) => {
            assert.equal(res.statusCode, 404);
            assert.equal(html.toString(), body);
            done();
          });
        });
      });

      it('serves a 500.html', done => {
        fs.readFile(join(__dirname, '..', 'src', 'public', 'default.html'), function(err, html) {
          request('http://localhost:5050/error', (error, res, body) => {
            assert.equal(res.statusCode, 500);
            assert.equal(html.toString(), body);
            done();
          });
        });
      });
    });

    describe('application/json format', () => {
      it('404', done => {
        request({
          url: 'http://localhost:5050/error',
          json: true
        }, (error, res, body) => {
          assert.equal(res.statusCode, 500);
          assert.deepEqual(body, {
            name: 'GeneralError',
            message: 'Something went wrong',
            code: 500,
            className: 'general-error',
            data: {},
            errors: {}
          });
          done();
        });
      });

      it('500', done => {
        request({
          url: 'http://localhost:5050/path/to/nowhere',
          json: true
        }, (error, res, body) => {
          assert.equal(res.statusCode, 404);
          assert.deepEqual(body, { name: 'NotFound',
            message: 'File not found',
            code: 404,
            className: 'not-found',
            errors: {}
          });
          done();
        });
      });
    });
  });
});


lib/errors.js
/*
 * feathers-errors
 * https://github.com/feathersjs/feathers-errors
 *
 * Copyright (c) 2014 Eric Kryski
 * Licensed under the MIT license.
 */

'use strict';

var _ = require('lodash');
var fs = require('fs');
var path = require('path');
var errors = require('./error-types');
var html = fs.readFileSync(path.resolve(__dirname, '..', 'public/error.html')).toString();

/**
* Escape the given string of `html`.
*
* @param {String} html
* @return {String}
* @api private
*/
var escapeHTML = function(html){
  return String(html)
    .replace(/&(?!\w+;)/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
};

/**
 * Method for a 404 middleware
 * See http://expressjs.com/guide.html#error-handling
 * @param  {Error} err - An error
 * @param  {Object} req - the request object
 * @param  {Object} res - the response object
 * @param  {Function} next - callback to call for next step in middleware chain
 */
var fourOhFour = function(req, res, next) {
  next(new errors.NotFound('Page not found.'));
};

/**
 * The error handler middleware.
 * See http://expressjs.com/guide.html#error-handling
 * @param  {Error} err - An error
 * @param  {Object} req - the request object
 * @param  {Object} res - the response object
 * @param  {Function} next - callback to call for next step in middleware chain
 *
 */

/* jshint unused:false */
var handler = function(err, req, res, next) {
  if (typeof err === 'string') {
    err = new errors.GeneralError(err);
  }
  else if (!(err instanceof errors.AbstractError)) {
    var oldError = err;
    err = new errors.GeneralError(oldError.message);

    if (oldError.stack) {
      err.stack = oldError.stack;
    }
  }

  var statusCode = typeof err.code === 'number' ? err.code : 500;

  res.status(statusCode);

  if (app.res.logger && typeof req.app.logger.error === 'function') {
      req.app.logger.error(req.url, err.stack || err);
  }
  else if (typeof req.app.error === 'function') {
    req.app.error(req.url, err.stack || err);
  }

  res.format({
    'text/html': function(){
      // If we have a rendering engine don't show the
      // default feathers error page.
      // TODO (EK): We should let people specify this instead
      // of making a shitty assumption.
      if (req.app.get('view engine') !== undefined) {
        if (err.code === 404) {
          return res.redirect('/404');
        }

        return res.redirect('/500');
      }

      var stack = (err.stack || '')
          .split('\n')
          .slice(1)
          .map(function(v) {
             return '<li>' + v + '</li>';
           })
          .join('');

      var errorPage = html
          .replace('{stack}', stack)
          .replace('{title}', err.message)
          .replace('{statusCode}', err.code)
          .replace(/\{error\}/g, escapeHTML(err.toString().replace(/\n/g, '<br/>')));
      res.send(errorPage);
    },

    'application/json': function(){
      res.json({
        'code': err.code,
        'name': err.name,
        'message': err.message,
        'errors': err.errors || {}
      });
    },

    'text/plain': function(){
      res.send(err.message);
    }
  });
};

exports = module.exports = function (config) {
  if (typeof config === 'function') {
    handler = config;
    config = {};
  }

  config = _.defaults({}, config, {
    handler: handler,
    fourOhFour: fourOhFour
  });

  return function () {
    var app = this;

    // Enable the errors Plugin
    app.enable('feathers errors');

    // Set the error handlers
    app.use(config.fourOhFour).use(config.handler);

    // Set the available errors on the app for convenience
    app.errors = errors;
  };
};

exports.fourOhFour = fourOhFour;
exports.handler = handler;
exports.types = errors;

src/error-handler.js
// jshint unused:false
import path from 'path';
import errors from './index';

const defaults = {
  public: './public'
};

export default function(options = {}) {
  options = Object.assign({}, defaults, options);

  return function(error, req, res, next) {
    if ( !(error instanceof errors.FeathersError) ) {
      let oldError = error;
      error = new errors.GeneralError(oldError.message, {errors: oldError.errors});

      if (oldError.stack) {
        error.stack = oldError.stack;
      }
    }

    const code = !isNaN( parseInt(error.code, 10) ) ? parseInt(error.code, 10) : 500;

    // Don't show stack trace if it is a 404 error
    if (code === 404) {
      error.stack = null;
    }

    res.status(code);

    res.format({
      'text/html': function() {
        const file = code === 404 ? '404.html' : '500.html';
        res.sendFile(path.join(options.public, file));
      },

      'application/json': function () {
        let output = Object.assign({}, error.toJSON());

        if (process.env.NODE_ENV === 'production') {
          delete output.stack;
        }

        res.json(output);
      }
    });
  };
}


src/index.js
const debug = require('debug')('feathers-errors');

// NOTE (EK): Babel doesn't properly support extending
// some classes in ES6. The Error class being one of them.
// Node v5.0+ does support this but until we want to drop support
// for older versions we need this hack.
// http://stackoverflow.com/questions/33870684/why-doesnt-instanceof-work-on-instances-of-error-subclasses-under-babel-node
// https://github.com/loganfsmyth/babel-plugin-transform-builtin-extend

class FeathersError extends Error {
  constructor (msg, name, code, className, data) {
    msg = msg || 'Error';

    let errors;
    let message;

    if (msg instanceof Error) {
      message = msg.message || 'Error';

      // NOTE (EK): This is typically to handle validation errors
      if (msg.errors) {
        errors = msg.errors;
      }
    } else if (typeof msg === 'object') { // Support plain old objects
      message = msg.message || 'Error';
      data = msg;
    } else { // message is just a string
      message = msg;
    }

    if (data && data.errors) {
      data = Object.assign({}, data);
      errors = data.errors;
      delete data.errors;
    }

    super(message);

    // NOTE (EK): Babel doesn't support this so
    // we have to pass in the class name manually.
    // this.name = this.constructor.name;
    this.type = 'FeathersError';
    this.name = name;
    this.message = message;
    this.code = code;
    this.className = className;
    this.data = data;
    this.errors = errors || {};

    debug(`${this.name}(${this.code}): ${this.message}`);
  }

  // NOTE (EK): A little hack to get around `message` not
  // being included in the default toJSON call.
  toJSON () {
    return {
      name: this.name,
      message: this.message,
      code: this.code,
      className: this.className,
      data: this.data,
      errors: this.errors
    };
  }
}

class BadRequest extends FeathersError {
  constructor (message, data) {
    super(message, 'BadRequest', 400, 'bad-request', data);
  }
}

class NotAuthenticated extends FeathersError {
  constructor (message, data) {
    super(message, 'NotAuthenticated', 401, 'not-authenticated', data);
  }
}

class PaymentError extends FeathersError {
  constructor (message, data) {
    super(message, 'PaymentError', 402, 'payment-error', data);
  }
}

class Forbidden extends FeathersError {
  constructor (message, data) {
    super(message, 'Forbidden', 403, 'forbidden', data);
  }
}

class NotFound extends FeathersError {
  constructor (message, data) {
    super(message, 'NotFound', 404, 'not-found', data);
  }
}

class MethodNotAllowed extends FeathersError {
  constructor (message, data) {
    super(message, 'MethodNotAllowed', 405, 'method-not-allowed', data);
  }
}

class NotAcceptable extends FeathersError {
  constructor (message, data) {
    super(message, 'NotAcceptable', 406, 'not-acceptable', data);
  }
}

class Timeout extends FeathersError {
  constructor (message, data) {
    super(message, 'Timeout', 408, 'timeout', data);
  }
}

class Conflict extends FeathersError {
  constructor (message, data) {
    super(message, 'Conflict', 409, 'conflict', data);
  }
}

class LengthRequired extends FeathersError {
  constructor (message, data) {
    super(message, 'LengthRequired', 411, 'length-required', data);
  }
}

class Unprocessable extends FeathersError {
  constructor (message, data) {
    super(message, 'Unprocessable', 422, 'unprocessable', data);
  }
}

class TooManyRequests extends FeathersError {
  constructor (message, data) {
    super(message, 'TooManyRequests', 429, 'too-many-requests', data);
  }
}

class GeneralError extends FeathersError {
  constructor (message, data) {
    super(message, 'GeneralError', 500, 'general-error', data);
  }
}

class NotImplemented extends FeathersError {
  constructor (message, data) {
    super(message, 'NotImplemented', 501, 'not-implemented', data);
  }
}

class BadGateway extends FeathersError {
  constructor (message, data) {
    super(message, 'BadGateway', 502, 'bad-gateway', data);
  }
}

class Unavailable extends FeathersError {
  constructor (message, data) {
    super(message, 'Unavailable', 503, 'unavailable', data);
  }
}

const errors = {
  FeathersError,
  BadRequest,
  NotAuthenticated,
  PaymentError,
  Forbidden,
  NotFound,
  MethodNotAllowed,
  NotAcceptable,
  Timeout,
  Conflict,
  LengthRequired,
  Unprocessable,
  TooManyRequests,
  GeneralError,
  NotImplemented,
  BadGateway,
  Unavailable,
  400: BadRequest,
  401: NotAuthenticated,
  402: PaymentError,
  403: Forbidden,
  404: NotFound,
  405: MethodNotAllowed,
  406: NotAcceptable,
  408: Timeout,
  409: Conflict,
  411: LengthRequired,
  422: Unprocessable,
  429: TooManyRequests,
  500: GeneralError,
  501: NotImplemented,
  502: BadGateway,
  503: Unavailable
};

function convert (error) {
  if (!error) {
    return error;
  }

  const FeathersError = errors[error.name];
  const result = FeathersError ? new FeathersError(error.message, error.data)
    : new Error(error.message || error);

  if (typeof error === 'object') {
    Object.assign(result, error);
  }

  return result;
}

export default Object.assign({
  convert,
  types: errors,
  errors
}, errors);
